File: app-final/constructs/kubernetes-web-app.ts
import { Construct } from "constructs";
import * as kubernetes from "../.gen/providers/kubernetes";
import { TerraformOutput } from "cdktf";

export interface KubernetesWebAppDeploymentConfig {
  readonly image: string;
  readonly replicas: string;
  readonly appName: string;
  readonly environment: string;
};

export interface KubernetesNodePortServiceConfig {
    readonly port: number;
    readonly appName: string;
    readonly environment: string;
};

export class SimpleKubernetesWebApp extends Construct {
    public readonly deployment: KubernetesWebAppDeployment;
    public readonly service: KubernetesNodePortService;
    public readonly config: KubernetesWebAppDeploymentConfig & KubernetesNodePortServiceConfig;

    constructor(scope: Construct, name: string, config: KubernetesWebAppDeploymentConfig & KubernetesNodePortServiceConfig) {
        super(scope, name);

        this.config = config;
        this.deployment = new KubernetesWebAppDeployment(this, name + "-deployment",
        {image: config.image, replicas: config.replicas, appName: config.appName, environment: config.environment}
        );

        this.service = new KubernetesNodePortService(this, name + "-service",
          {port: config.port, appName: config.appName, environment: config.environment}
        );

        new TerraformOutput(this, "frontend_url", {
            value: "http://localhost:" + config.port.toString(),
          });
      
    }
};

export class KubernetesWebAppDeployment extends Construct {
  public readonly resource: kubernetes.Deployment;

  constructor(scope: Construct, name: string, config: KubernetesWebAppDeploymentConfig) {
    super(scope, name);

    this.resource = new kubernetes.Deployment(this, name, {
    metadata: [
        {
        labels:
            {
            app: config.appName,
            environment: config.environment,
            },
        name: config.appName,
        },
    ],
    spec: [
        {
        replicas: config.replicas,
        selector: [
            {
            matchLabels:
                {
                environment: config.environment,
                app: config.appName
                },
            },
        ],
        template: [
            {
            metadata: [
                {
                labels: {
                    app: config.appName,
                    environment: config.environment
                },
                },
            ],
            spec: [
                {
                container: [
                    {
                    image: config.image,
                    name: config.appName,
                    },
                ],
                },
            ],
            },
        ],
        },
    ],
    });
}
};

export class KubernetesNodePortService extends Construct {
    public readonly resource: kubernetes.Service;

    constructor(scope: Construct, name: string, config: KubernetesNodePortServiceConfig) {
    super(scope, name);

    this.resource = new kubernetes.Service(this, name, {
        metadata: [
        {
            name: config.appName,
        },
        ],
        spec: [
        {
            type: "NodePort",
            port: [
            {
                port: 80,
                targetPort: "80",
                nodePort: config.port,
                protocol: "TCP"
            },
            ],
            selector: {
            app: config.appName,
            },
        },
        ],
    });
    };
};    

File: cdktf-examples/java/src/main/java/com/mycompany/constructs/KubernetesWebApp.java
package com.mycompany.constructs;

import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Map.Entry;

import com.hashicorp.cdktf.providers.kubernetes.deployment.*;
import software.constructs.Construct;

public class KubernetesWebApp extends Construct {
    public Deployment resource;

    public KubernetesWebApp(Construct scope, String name, KubernetesWebAppDeploymentConfig config) {
        super(scope, name);

        String deploymentName = String.format("%s-%s-%s", config.app(), config.component(), config.environment());

        List<DeploymentSpecTemplateSpecContainerEnv> env;
        if (config.env() != null) {
            Set<Entry<String, String>> entries = config.env().entrySet();
            env = new java.util.ArrayList<>(entries.size());

            for (Entry<String, String> entry : entries) {
                env.add(DeploymentSpecTemplateSpecContainerEnv.builder()
                        .name(entry.getKey())
                        .value(entry.getValue())
                        .build());
            }
        } else {
            env = new java.util.ArrayList<DeploymentSpecTemplateSpecContainerEnv>(0);
        }

        this.resource = new Deployment(this, name, DeploymentConfig.builder()
                .metadata(DeploymentMetadata.builder()
                        .labels(Map.of(
                                "app", config.app(),
                                "component", config.component(),
                                "environment", config.environment()))
                        .name(deploymentName)
                        .build())
                .spec(DeploymentSpec.builder()
                        .replicas(config.replicas().toString())
                        .selector(DeploymentSpecSelector.builder()
                                .matchLabels(Map.of(
                                        "app", config.app(),
                                        "component", config.component(),
                                        "environment", config.environment()))
                                .build())
                        .template(DeploymentSpecTemplate.builder()
                                .metadata(DeploymentSpecTemplateMetadata.builder()
                                        .labels(Map.of(
                                                "app", config.app(),
                                                "component", config.component(),
                                                "environment", config.environment()))
                                        .build())
                                .spec(DeploymentSpecTemplateSpec.builder()
                                        .container(List.of(
                                                DeploymentSpecTemplateSpecContainer.builder()
                                                        .image(config.image())
                                                        .name(deploymentName)
                                                        .env(env)
                                                        .build()))
                                        .build())
                                .build())
                        .build())
                .build());
    }
}
